src/module_builder.rs

```diff
@@ -0,0 +1,129 @@
+use std::collections::HashSet;
+
+use rquickjs::{loader::{BuiltinResolver, ModuleLoader, Resolver}, module::ModuleDef, Ctx, Result};
+use crate::{modules::{
+    buffer::BufferModule,
+    child_process::ChildProcessModule,
+    console::ConsoleModule,
+    crypto::CryptoModule,
+    encoding::HexModule,
+    events::EventsModule,
+    fs::{FsModule, FsPromisesModule},
+    module::ModuleModule,
+    navigator::NavigatorModule,
+    net::NetModule,
+    os::OsModule,
+    path::PathModule,
+    performance::PerformanceModule,
+    process::ProcessModule,
+    timers::TimersModule,
+    url::UrlModule,
+    uuid::UuidModule,
+    xml::XmlModule,
+}, utils::UtilModule};
+
+#[derive(Debug, Default)]
+pub struct ModuleResolver {
+    builtin_resolver: BuiltinResolver,
+}
+
+impl ModuleResolver {
+    #[must_use]
+    pub fn with_module<P: Into<String>>(mut self, path: P) -> Self {
+        self.builtin_resolver.add_module(path.into());
+        self
+    }
+}
+
+impl Resolver for ModuleResolver {
+    fn resolve(&mut self, ctx: &Ctx<'_>, base: &str, name: &str) -> Result<String> {
+        // Strip node prefix so that we support both with and without
+        let name = name.strip_prefix("node:").unwrap_or(name);
+
+        self.builtin_resolver.resolve(ctx, base, name)
+    }
+}
+
+struct ModuleInfo<T: ModuleDef> {
+    name: &'static str,
+    module: T,
+}
+
+pub struct ModuleBuilder {
+    builtin_resolver: ModuleResolver,
+    module_loader: ModuleLoader,
+    module_names: HashSet<&'static str>,
+    init_global: Vec<fn(&Ctx<'_>) -> Result<()>>
+}
+
+impl ModuleBuilder {
+    pub fn new() -> Self {
+        Self {
+            builtin_resolver: ModuleResolver::default(),
+            module_loader: ModuleLoader::default(),
+            module_names: HashSet::new(),
+            init_global: Vec::new(),
+        }
+    }
+
+    pub fn with_default() -> Self {
+        Self::new()
+            .with_module("crypto", CryptoModule)
```

Comment:
I like this approach! Is there a way to avoid hardcoding names of modules when using the builder pattern? Maybe an enum? Then we we can may implement `Into<ModuleInfo>` for the enum and use that as arguments for withModule?

Reply:
I thought of something like that, but how would people add their own modules without changing the internal enum?
I actually wanted to add a name field in the module definitions, but it wasn't object safe so gave up on that.

Reply:
They would just implement` Into<ModuleInfo>` for their structs or enum or whatever wrapper they use

Reply:
All names have been moved to the modules themself with a into :)